core/drivers/pl022_spi.c

```diff
@@ -0,0 +1,532 @@
+		DMSG("speed: requested: %u, closest1: %u\n",
+			pd->speed_hz, freq1);
+	} else {
+		*cpsdvr = tmp_cpsdvr2;
+		*scr = tmp_scr2;
+		DMSG("speed: requested: %u, closest2: %u\n",
+			pd->speed_hz, freq2);
+	}
+	DMSG("cpsdvr: %u (0x%x), scr: %u (0x%x)\n",
+		*cpsdvr, *cpsdvr, *scr, *scr);
+}
+
+static void pl022_flush_fifo(struct pl022_data *pd)
+{
+	uint32_t rdat;
+
+	do {
+		while (read32(pd->base + SSPSR) & SSPSR_RNE) {
+			rdat = read32(pd->base + SSPDR);
+			DMSG("rdat: 0x%x\n", rdat);
+		}
+	} while (read32(pd->base + SSPSR) & SSPSR_BSY);
+}
+
+void pl022_configure(struct pl022_data *pd)
+{
+	uint16_t mode, data_size;
+	uint8_t cpsdvr, scr, lbm;
+
+	pl022_sanity_check(pd);
+	pl022_calc_clk_divisors(pd, &cpsdvr, &scr);
+
+	/* configure ssp based on platform settings */
+	switch (pd->mode) {
+	case SPI_MODE0:
+		DMSG("SPI_MODE0\n");
+		mode = PL022_SPI_MODE0;
+		break;
+	case SPI_MODE1:
+		DMSG("SPI_MODE1\n");
+		mode = PL022_SPI_MODE1;
+		break;
+	case SPI_MODE2:
+		DMSG("SPI_MODE2\n");
+		mode = PL022_SPI_MODE2;
+		break;
+	case SPI_MODE3:
+		DMSG("SPI_MODE3\n");
+		mode = PL022_SPI_MODE3;
+		break;
+	default:
+		EMSG("Invalid spi mode: %u\n", pd->mode);
+		return;
+	}
+
+	switch (pd->data_size_bits) {
+	case 8:
+		DMSG("Data size: 8\n");
+		data_size = PL022_DATA_SIZE8;
+		break;
+	case 16:
+		DMSG("Data size: 16\n");
+		data_size = PL022_DATA_SIZE16;
+		break;
+	default:
+		EMSG("Unsupported data size: %u bits\n", pd->data_size_bits);
+		return;
+	}
+
+	if (pd->loopback) {
+		DMSG("Starting in loopback mode!\n");
+		lbm = SSPCR1_LBM_YES;
+	} else {
+		DMSG("Starting in regular (non-loopback) mode!\n");
+		lbm = SSPCR1_LBM_NO;
+	}
+
+	DMSG("set serial clock rate (scr), spi mode (phase and clock)\n");
+	DMSG("set frame format (spi) and data size (8- or 16-bit)\n");
+	io_mask16(pd->base + SSPCR0, SHIFT_U32(scr, 8) | mode | SSPCR0_FRF_SPI
+		| data_size, MASK_16);
+
+	DMSG("set master mode, disable ssp, set loopback mode\n");
+	io_mask8(pd->base + SSPCR1, SSPCR1_SOD_DISABLE | SSPCR1_MS_MASTER
+		| SSPCR1_SSE_DISABLE | lbm, MASK_4);
+
+	DMSG("set clock prescale\n");
+	io_mask8(pd->base + SSPCPSR, cpsdvr, SSPCPSR_CPSDVR);
+
+	DMSG("disable interrupts\n");
+	io_mask8(pd->base + SSPIMSC, 0, MASK_4);
+
+	DMSG("set cs gpio dir to out\n");
+	gpio_set_direction(pd->cs_gpio_pin, GPIO_DIR_OUT);
+
+	DMSG("pull cs high\n");
+	gpio_set_value(pd->cs_gpio_pin, GPIO_LEVEL_HIGH);
+}
+
+void pl022_start(struct pl022_data *pd)
+{
+	DMSG("empty FIFO before starting");
+	pl022_flush_fifo(pd);
+
+	DMSG("enable SSP");
+	io_mask8(pd->base + SSPCR1, SSPCR1_SSE_ENABLE, SSPCR1_SSE);
+}
+
+void pl022_end(struct pl022_data *pd)
+{
+	/* disable ssp */
+	io_mask8(pd->base + SSPCR1, SSPCR1_SSE_DISABLE, SSPCR1_SSE);
+}
+
+static const struct spi_ops pl022_ops = {
+	.txrx8 = pl022_txrx8,
+	.txrx16 = pl022_txrx16,
+	.tx8 = pl022_tx8,
+	.tx16 = pl022_tx16,
+	.rx8 = pl022_rx8,
+	.rx16 = pl022_rx16,
+};
+
+void pl022_init(struct pl022_data *pd)
+{
+	pd->chip.ops = &pl022_ops;
```

Comment:
ok but warn caller that pl022_data must be kzalloced or at least memset to 0 before being filled.

a good thing could be `pl022_init` to alloc & reset `pl022_data` structure and init the internal stuff (as `chip.ops`) and return the data point.  A `pl022_release()` would be needed to properly free the structure and its resources (if any) but this might be just dead code.